### DISCUSSION TITLE
Create an attribute that, when applied to a struct type, indicates that the default value of that struct is invalid

### DISCUSSION THREAD

**Discussion by Xenoprimate** [dotnet/roslyn#70600] [Ideas]
With the latest versions of C# it is possible to override the default ctor for struct types which is useful to help prevent badly-formed struct-type values, but does not prevent usage of the default(T) value of a struct-type (either explicity or implicitly e.g. as part of an array or non-initialized class member).

Although the compiler can never fully prevent someone from creating a value of default(T), having an attribute we can apply to struct types that gives a hint would be useful. In these cases, where the compiler detects that a value will be default(T) for a type T that is annotated with this proposed attribute, it can emit a warning.

The compiler would only have to detect the most obvious cases-- but these would still be very useful in minimizing errors.

Examples always help, so here's an example:

```
[DefaultValueIsInvalid] // Proposed attribute
readonly struct UnitQuaternion {
    public readonly float X;
    public readonly float Y;
    public readonly float Z;
    public readonly float W;
    // etc...
}

// Then, later:

static void Test() {
    var uqArr = new UnitQuaternion[4]; // Warning here unless we Array.Fill?
    DoSomething(uqArr[0]); // Or warning here that we're probably using an invalid value?
}

class SomeClass {
    UnitQuaternion _uq; // Warning here that _uq will be left at default(T)?
}
```

**Comment by CyrusNajmabadi** [comment]
Closing out.  This is already possible today with analyzers.  Just define that attribute, and write an analzyer that does whatever detection you want.  

> The compiler would only have to detect the most obvious cases-- but these would still be very useful in minimizing errors.

The benefit of doing your own analysis is that you can define exactly what the 'obvious cases' are and expand on htat later, especially with domain knowledge.
